packages/pwa-buildpack/lib/WebpackTools/plugins/RootComponentsPlugin.js

```diff
@@ -0,0 +1,217 @@
+                            'found %s directives: %o',
+                            rootComponentDirectives.length,
+                            rootComponentDirectives
+                        );
+                        if (rootComponentDirectives.length === 0) {
+                            return;
+                        }
+
+                        if (rootComponentDirectives.length > 1) {
+                            console.warn(
+                                `Found more than 1 RootComponent Directive in ${rootComponentFile}. Only the first will be used`
+                            );
+                        }
+
+                        const {
+                            pageTypes,
+                            variant
+                        } = rootComponentDirectives[0];
+
+                        if (!pageTypes || pageTypes.length === 0) {
+                            console.warn(
+                                `No pageTypes specified for RootComponent ${rootComponentFile}. RootComponent will never be used.`
+                            );
+                        } else {
+                            pageTypes.forEach(pageType => {
+                                debug(
+                                    'creating root component map key for %s',
+                                    rootComponentFile,
+                                    pageType
+                                );
+                                const key = toRootComponentMapKey(
+                                    pageType,
+                                    variant
+                                );
+                                importerSources[
+                                    key
+                                ] = `() => import(/* webpackChunkName: "${key}" */'${relative(
+                                    context,
+                                    rootComponentFile
+                                )}')`;
+                            });
+                        }
+                    })
+                );
+                return importerSources;
+            },
+            Promise.resolve({})
+        );
+
+        const rootComponentEntries = Object.entries(rootComponentImporters);
+
+        if (rootComponentEntries.length === 0) {
+            prettyLogger.error(
+                `No RootComponents were found in any of these directories: \n - ${rootComponentsDirsAbs.join(
+                    '\n - '
+                )}.\n\nThe MagentoRouter will not be able to load SEO URLs.`
+            );
+        }
+
+        this.contents = `
+            const rootComponentsMap = {
+            ${rootComponentEntries.map(entry => entry.join(':')).join(',\n')}
+            };
+            const key = ${toRootComponentMapKey.toString()};
+            export default function fetchRootComponent(type, variant = 'default') {
+                return rootComponentsMap[key(type, variant)]().then(m => m.default || m);
```

Comment:
Just for my sanity, how can I view this output unminified? If I run `NODE_ENV=development yarn build` the output `dist/client.js` is still minified/uglified.

Reply:
This is kinda confusing config, isn't it. In my mind, "developer mode" meant hot reloading mode. Different bundle splitting behaviors, `webpack-dev-server` instead of `webpack`, et cetera. But it makes a lot of sense to just collapse this variable into developer mode, and keep the same bundle splitting rules. I think the only reason we didn't was for hot reload considerations, but it needs a revisit.